cmd/frontend/graphqlbackend/user_collaborators.go

```diff
@@ -0,0 +1,111 @@
+package graphqlbackend
+
+import (
+	"context"
+
+	"github.com/inconshreveable/log15"
+
+	"sync"
+
+	"github.com/sourcegraph/sourcegraph/cmd/frontend/backend"
+	"github.com/sourcegraph/sourcegraph/internal/actor"
+	"github.com/sourcegraph/sourcegraph/internal/api"
+	"github.com/sourcegraph/sourcegraph/internal/authz"
+	"github.com/sourcegraph/sourcegraph/internal/database"
+	"github.com/sourcegraph/sourcegraph/internal/gitserver/gitdomain"
+	"github.com/sourcegraph/sourcegraph/internal/goroutine"
+	"github.com/sourcegraph/sourcegraph/internal/types"
+	"github.com/sourcegraph/sourcegraph/internal/vcs/git"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+func (r *UserResolver) InvitableCollaborators(ctx context.Context) ([]*invitableCollaboratorResolver, error) {
+	a := actor.FromContext(ctx)
+	if !a.IsAuthenticated() {
+		return nil, errors.New("no current user")
+	}
+
+	// We'll search for collaborators in 25 of the user's most-starred repositories.
+	const maxReposToScan = 25
+	pickedRepos, err := backend.NewRepos(r.db.Repos()).List(ctx, database.ReposListOptions{
+		// SECURITY: This must be the authenticated user's ID.
+		UserID:                 a.UID,
+		IncludeUserPublicRepos: true,
+		OrderBy: database.RepoListOrderBy{{
+			Field:      "stars",
+			Descending: true,
+		}},
+		LimitOffset: &database.LimitOffset{Limit: maxReposToScan},
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "Repos.List")
+	}
+
+	// In parallel collect all recent committers info for the few repos we're going to scan.
+	recentCommitters, err := gitserverParallelRecentCommitters(ctx, pickedRepos, git.Commits)
+	if err != nil {
+		return nil, err
+	}
+
+	authUserEmails, err := database.UserEmails(r.db).ListByUser(ctx, database.UserEmailsListOptions{
+		UserID: a.UID,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	userExists := func(username, email string) bool {
+		if username != "" {
+			_, err := r.db.Users().GetByUsername(ctx, username)
+			return err == nil
+		}
+		_, err := r.db.Users().GetByVerifiedEmail(ctx, email)
+		return err == nil
```

Comment:
If I understand correctly, [the userExists function will be invoked once with username being an empty string and once with email being an empty string](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/frontend/graphqlbackend/external_service_collaborators.go?L239-247).

However, in our example below we make the username _always_ an empty string so what happens is we run the second part of the query here twice (`r.db.Users().GetByVerifiedEmail(ctx, email)`). Ones with email being an empty string. And that for every result that we find!

I think we should probably guard the second query as well:

```suggestion
		if email != "" {
		  _, err := r.db.Users().GetByVerifiedEmail(ctx, email)
		  return err == nil
		}
		return false
```

Reply:
I've clarified this logic, splitting `userExists` into two variants `userExistsByUsername` and `userExistsByEmail` to make this easier to reason about. I've also guarded against empty usernames effectivey now.